### PR TITLE
Disable client-initiated renegotiation by default

### PR DESCRIPTION
**Pull request by richsalz**
This replaces #15051 which I will close.

This PR is *slightly* bigger but it is much more clean and easy to understand.

### CODE REVIEW

File: include/openssl/ssl.h.in [hunk 1]
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
...
-/* Enable support for Kernel TLS */
+    /* Enable support for Kernel TLS */
 # define SSL_OP_ENABLE_KTLS                              0x00000008U
-
 # define SSL_OP_TLSEXT_PADDING                           0x00000010U
+# define SSL_OP__UNUXED                                  0x00000020U
 # define SSL_OP_SAFARI_ECDHE_ECDSA_BUG                   0x00000040U
 # define SSL_OP_IGNORE_UNEXPECTED_EOF                    0x00000080U
-
+    /* See below SSL_OP_ALLOW_CLIENT_RENEGOTIATION           0x00000100U */
```
Comment (mattcaswell) [inline include/openssl/ssl.h.in:335]: Why is this one out of order? I think it should be defined here in this list

File: include/openssl/ssl.h.in [hunk 2]
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
...
-/* Allow initial connection to servers that don't support RI */
+    /* Allow initial connection to servers that don't support RI */
 # define SSL_OP_LEGACY_SERVER_CONNECT                    0x00000004U
-
-/* Enable support for Kernel TLS */
+    /* Enable support for Kernel TLS */
 # define SSL_OP_ENABLE_KTLS                              0x00000008U
-
 # define SSL_OP_TLSEXT_PADDING                           0x00000010U
+# define SSL_OP__UNUXED                                  0x00000020U
```
Comment (mattcaswell) [inline include/openssl/ssl.h.in:332]: I think this should just be a comment noting that it is unused. Adding a special define means that people might start relying on it being there. If we end up using a `uint64_t` for the OP value then there will be lots of unused slots and it doesn't seem sensible to have a define for each one.

File: test/helpers/ssltestlib.c [hunk 3]
```diff
@@ -695,7 +695,9 @@ int create_ssl_ctx_pair(OSSL_LIB_CTX *libctx, const SSL_METHOD *sm,
     if (sctx != NULL) {
         if (*sctx != NULL)
             serverctx = *sctx;
-        else if (!TEST_ptr(serverctx = SSL_CTX_new_ex(libctx, NULL, sm)))
+        else if (!TEST_ptr(serverctx = SSL_CTX_new_ex(libctx, NULL, sm))
+            || !TEST_true(SSL_CTX_set_options(serverctx,
+                    SSL_OP_ALLOW_CLIENT_RENEGOTIATION)))
```
Comment (mattcaswell) [inline test/helpers/ssltestlib.c:700]: Indentation seems odd here

File: ssl/ssl_lib.c [hunk 4]
```diff
@@ -2277,39 +2277,47 @@ int SSL_get_key_update_type(const SSL *s)
...
 {
     if (SSL_IS_TLS13(s)) {
         ERR_raise(ERR_LIB_SSL, SSL_R_WRONG_SSL_VERSION);
         return 0;
     }
 
-    if ((s->options & SSL_OP_NO_RENEGOTIATION)) {
-        ERR_raise(ERR_LIB_SSL, SSL_R_NO_RENEGOTIATION);
+    if (SSL_is_server(s)
+            && (s->options & SSL_OP_ALLOW_CLIENT_RENEGOTIATION) == 0)
```
Comment (mattcaswell) [inline ssl/ssl_lib.c:2292]: This seems wrong. It appears to prevent server initiated renegotiation (on the server side) if the option is not set.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I was expecting a change in statem_srvr.c here:

https://github.com/openssl/openssl/blob/22d1138fe2fde9a16e80b81de1d848ae6fa879ef/ssl/statem/statem_srvr.c#L1378-L1393

The changes in the tests seem to be mainly about fixing the tests that were broken by this. There doesn't seem to be a test to check that client initiated reneg doesn't work by default but server initiated reneg does.

File: ssl/ssl_lib.c (hunk 4 above)
Comment (richsalz) [reply ssl/ssl_lib.c:2292]: Yes, this is in the wrong place. It should be, as you said, around line 1384 of `statem_server.c`

I'll address all your feedback by Monday; today's a holiday/wellness-day for work.

File: include/openssl/ssl.h.in (hunk 1 above)
Comment (richsalz) [reply include/openssl/ssl.h.in:335]: I thought it more useful to keep related options together, but the current `0x000....U` syntax makes that hard. That's why I'd like to use the `1<<xxx` syntax. But that's discussed in #15145 so I moved it.

**Comment by mattcaswell** [comment]
> pushed an updated commit

Did you really? I don't see it...

> Looks like both of those involve the TLSProxy, about which I know nothing. Any clues appreciated while I start reading that code.

My guess is something like this might help:

```` diff
diff --git a/test/recipes/70-test_renegotiation.t b/test/recipes/70-test_renegotiation.t
index 256964266c..1cf14a444e 100644
--- a/test/recipes/70-test_renegotiation.t
+++ b/test/recipes/70-test_renegotiation.t
@@ -36,6 +36,7 @@ my $proxy = TLSProxy::Proxy->new(
 
 #Test 1: A basic renegotiation test
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start() or plan skip_all => "Unable to start up Proxy for tests";
 plan tests => 4;
@@ -45,6 +46,7 @@ ok(TLSProxy::Message->success(), "Basic renegotiation");
 $proxy->clear();
 $proxy->filter(\&reneg_filter);
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start();
 ok(TLSProxy::Message->fail(), "No client SCSV");
@@ -58,7 +60,7 @@ SKIP: {
     $proxy->filter(undef);
     $proxy->ciphers("DEFAULT:\@SECLEVEL=0");
     $proxy->clientflags("-no_tls1_3 -cipher AES128-SHA:\@SECLEVEL=0");
-    $proxy->serverflags("-no_tls1_3 -no_tls1_2");
+    $proxy->serverflags("-no_tls1_3 -no_tls1_2 -client_renegotiation");
     $proxy->reneg(1);
     $proxy->start();
     my $chversion;
@@ -87,6 +89,7 @@ SKIP: {
     $proxy->clear();
     $proxy->filter(\&sigalgs_filter);
     $proxy->clientflags("-tls1_2");
+    $proxy->serverflags("-client_renegotiation");
     $proxy->reneg(1);
     $proxy->start();
     ok(TLSProxy::Message->fail(), "client_sig_algs instead of sig_algs");
diff --git a/test/recipes/70-test_sslmessages.t b/test/recipes/70-test_sslmessages.t
index 3f57af62d5..befc4c7e9e 100644
--- a/test/recipes/70-test_sslmessages.t
+++ b/test/recipes/70-test_sslmessages.t
@@ -239,6 +239,7 @@ checkhandshake($proxy, checkhandshake::CLIENT_AUTH_HANDSHAKE,
 #Test 7: A handshake with a renegotiation
 $proxy->clear();
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start();
 checkhandshake($proxy, checkhandshake::RENEG_HANDSHAKE,
````

**Comment by richsalz** [comment]
> Didn't see commit.

Oops.  Forgot to get return  :)

I applied your patch and get `Error with command: "-client_renegotiation"` which i think i can figure out.

**Comment by richsalz** [comment]
Thanks very much @mattcaswell!  I still need to write another two tests, right?

File: include/openssl/ssl.h.in [hunk 5]
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
...
+    /* Turn on Cookie Exchange (on relevant for servers) */
 # define SSL_OP_COOKIE_EXCHANGE                          0x00002000U
-/* Don't use RFC4507 ticket extension */
+    /* Don't use RFC4507 ticket extension */
 # define SSL_OP_NO_TICKET                                0x00004000U
 # ifndef OPENSSL_NO_DTLS1_METHOD
-/* Use Cisco's "speshul" version of DTLS_BAD_VER
- * (only with deprecated DTLSv1_client_method())  */
+    /*
+     * Use Cisco's "speshul" version of DTLS_BAD_VER
```
[the remark this answers is not in the document]
Comment (richsalz) [reply include/openssl/ssl.h.in]: Yes, modified.

**Comment by richsalz** [comment]
fixup commit pushed.  will work on adding the new tests now.

**Review by mattcaswell — APPROVED** [review]
I would have preferred not to have to add a new command line option just for a test - but ok.

File: apps/lib/s_cb.c [hunk 6]
```diff
@@ -1233,12 +1233,10 @@ int config_ctx(SSL_CONF_CTX *cctx, STACK_OF(OPENSSL_STRING) *str,
...
         const char *flag = sk_OPENSSL_STRING_value(str, i);
         const char *arg = sk_OPENSSL_STRING_value(str, i + 1);
+
         if (SSL_CONF_cmd(cctx, flag, arg) <= 0) {
-            if (arg != NULL)
-                BIO_printf(bio_err, "Error with command: \"%s %s\"\n",
-                           flag, arg);
-            else
-                BIO_printf(bio_err, "Error with command: \"%s\"\n", flag);
+            BIO_printf(bio_err, "Call to SSL_CONF_md(%s, %s) failed\n",
```
Comment (t8m) [inline apps/lib/s_cb.c:1238]: Typo: SSL_CONF_cmd

File: doc/man3/SSL_CONF_cmd.pod [hunk 7]
```diff
@@ -58,9 +58,15 @@ Use server and not client preference order when determining which cipher suite,
...
+=item B<-client_renegotiation>
+
+Allows servers to accept client-initiated renegotiation. Equivalent to
+setting B<SSL_OP_ALLOW_CLIENT_RENEGOTIATION>.
+Only used by servers.
+
 =item B<-legacyrenegotiation>
 
-permits the use of unsafe legacy renegotiation. Equivalent to setting
+Permits the use of unsafe legacy renegotiation. Equivalent to setting
```
Comment (t8m) [inline doc/man3/SSL_CONF_cmd.pod:69]: The uppercase of the first word is missing for other options.You might correct them all if you are fixing this particular one. But not required of course.

File: doc/man3/SSL_CONF_cmd.pod [hunk 8]
```diff
@@ -70,13 +76,19 @@ B<SSL_OP_NO_RENEGOTIATION>.
 
 =item B<-no_resumption_on_reneg>
 
-set SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION flag. Only used by servers.
+set B<SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION> flag. Only used by servers.
```
Comment (t8m) [inline doc/man3/SSL_CONF_cmd.pod:79]: For example here.

File: ssl/ssl_lib.c [hunk 9]
```diff
@@ -2277,39 +2277,41 @@ int SSL_get_key_update_type(const SSL *s)
     return s->key_update;
 }
 
-int SSL_renegotiate(SSL *s)
+/*
+ * Can we accept a renegotiation request?  If yes, set the flag and
+ * return 1 if yes. If not, raise error and return 0.
+ */
+static int can_renegotiation(SSL *s)
```
Comment (t8m) [inline ssl/ssl_lib.c:2284]: `can_renegotiate` would sound more English to me. :grin: 

Also, I'd rather keep the duplication of `s->renegotiate = 1;` and keep the function read-only to align with the name.

**Review by t8m — CHANGES_REQUESTED** [review]
Some nits

**Comment by richsalz** [comment]
> I would have preferred not to have to add a new command line option just for a test - but ok.

I could not think of any other way to do it.  Happy to create a followon PR if you have any flash of insight. I do think the flag could be useful for testing clients, but that's rationalization :)  And we do have a options in the list command that are used only for the `cmd-nits` target. :)

File: doc/man3/SSL_CONF_cmd.pod (hunk 7 above)
Comment (richsalz) [reply doc/man3/SSL_CONF_cmd.pod:69]: There were only two uppercase fixes needed.

Comment (t8m) [reply doc/man3/SSL_CONF_cmd.pod:69]: Still need to push changes?

**Comment by richsalz** [comment]
fixup commit pushed addressing @t8m's feedback. To  avoid extra mail I just marked the issues as resolved.

File: doc/man3/SSL_CONF_cmd.pod (hunk 7 above)
Comment (richsalz) [reply doc/man3/SSL_CONF_cmd.pod:69]: i waited for 'make test' to finish :)

**Review by t8m — APPROVED** [review]
@mattcaswell still OK?

**Review by mattcaswell — APPROVED** [review]
Still ok

**Comment by mattcaswell** [comment]
> I could not think of any other way to do it. Happy to create a followon PR if you have any flash of insight.

I would have done it in sslapitest.c rather than in a TLSProxy test.

**Comment by richsalz** [comment]
> I would have done it in sslapitest.c rather than in a TLSProxy test.

okay, I'll think about it.

**Comment by richsalz** [comment]
Can someone change the labels to remove the "pending" and show it's approved?  Thanks.

**Comment by mattcaswell** [comment]
Oops. Sorry. Done.

**Comment by mattcaswell** [comment]
> I would have done it in sslapitest.c rather than in a TLSProxy test.

Actually an even better place might be in:

https://github.com/openssl/openssl/blob/master/test/ssl-tests/17-renegotiate.cnf.in

**Comment by richsalz** [comment]
Thanks for the pointer.  I'll look at that, too, but post-alpha-freeze. :)

**Comment by mattcaswell** [comment]
Ah....there's a non-trivial rebase required due to the 64-bit op PR

**Comment by richsalz** [comment]
By design, the ssl.h.in conflict is easily handled since that PR had all of this PR's changes in it :)  Rebase coming up.

**Comment by richsalz** [comment]
Rebased, squashed, and pushed. As mentioned above, the conflcit in `ssl.h.in` was handled by just taking the master version, and `CHANGES.md` was trivial. No other changes were made.

**Comment by mattcaswell** [comment]
Thanks. @t8m, since this ended up being trivial after all I think we probably don't need to reset the timer for it? Do you agree?

**Comment by t8m** [comment]
Sure, merging just now.

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution!

**Pull request closed without merging**